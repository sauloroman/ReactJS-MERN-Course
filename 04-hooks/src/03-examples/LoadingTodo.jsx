export const LoadingTodo = () => {
  return (
    <p data-testid="loading" className="loading">Cargando...</p>
  )
}
