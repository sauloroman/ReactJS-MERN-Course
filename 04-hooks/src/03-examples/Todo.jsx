import { useLayoutEffect, useState, useRef } from "react"

export const Todo = ({ completed, id, title, userId }) => {
  
  const todoRef = useRef()
  const [boxSize, setBoxSize] = useState({ width: 0, height: 0 })
  
  useLayoutEffect(() => {
    const { height, width } = todoRef.current.getBoundingClientRect()
    setBoxSize({ width, height })
  }, [id])

  return (
    <>
      <div data-testid="todo" className={`todo ${completed && 'todo--completed'}`}>
        <p ref={todoRef}>Title: <span data-testid="todo-title">{title}</span></p>
        <p>id: <span data-testid="todo-id">{id}</span></p>
        <p>userId: <span data-testid="todo-userId">{userId}</span></p>
      </div>

      <code>{JSON.stringify( boxSize )}</code>
    </>
  )
}
