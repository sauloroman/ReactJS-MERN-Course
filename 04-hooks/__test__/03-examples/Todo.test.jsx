import { render, screen } from '@testing-library/react'
import { Todo } from '../../src/03-examples/Todo'

describe("Tests in Todo Component", () => {

  const todoTest = {
    id: 1,
    title: "Todo 1",
    userId: 1,
    completed: false,
  }

  test("It must match with snapshot", () => {
    const { container } = render(<Todo {...todoTest} />)
    expect( container ).toMatchSnapshot()
  })

  test("It should show the title, id, and userId of the todo", () => {
    render(<Todo {...todoTest} />)

    const titleEle = screen.getByTestId("todo-title")
    const idEle = screen.getByTestId("todo-id")
    const userIdEle = screen.getByTestId("todo-userId")

    expect( titleEle.textContent ).toBe( todoTest.title )    
    expect( idEle.textContent ).toBe( String(todoTest.id) )    
    expect( userIdEle.textContent ).toBe( String(todoTest.userId) )    

  })

  test('It should add or remove a class depeding on the todo status', () => {

    todoTest.completed = true

    render(<Todo {...todoTest} />)

    const todoEle = screen.getByTestId("todo")

    expect( todoEle.className ).toBe("todo todo--completed")

  })

})