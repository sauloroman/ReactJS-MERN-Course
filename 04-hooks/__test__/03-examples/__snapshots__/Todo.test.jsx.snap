// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in Todo Component It must match with snapshot 1`] = `
<div>
  <div
    class="todo false"
    data-testid="todo"
  >
    <p>
      Title: 
      <span
        data-testid="todo-title"
      >
        Todo 1
      </span>
    </p>
    <p>
      id: 
      <span
        data-testid="todo-id"
      >
        1
      </span>
    </p>
    <p>
      userId: 
      <span
        data-testid="todo-userId"
      >
        1
      </span>
    </p>
  </div>
  <code>
    {"width":0,"height":0}
  </code>
</div>
`;
